ccl/changefeedccl: TestChangefeedDataTTL failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/668162cc99e4f3198b663b1abfa51858eeb3ccb8

Parameters:
```
TAGS=
GOFLAGS=
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestChangefeedDataTTL PKG=github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1212391&tab=buildLog

```
I190401 05:35:38.545298 26532 sql/event_log.go:135  [n1,intExec=initializeClusterSecret] Event: "set_cluster_setting", target: 0, info: {SettingName:cluster.secret Value:0335c4a5-31c0-41a0-b4b2-c2c536b7b26f User:root}
I190401 05:35:38.551632 26495 sql/event_log.go:135  [n1,intExec=create-default-db] Event: "create_database", target: 50, info: {DatabaseName:defaultdb Statement:CREATE DATABASE IF NOT EXISTS defaultdb User:root}
I190401 05:35:38.554466 26563 sql/event_log.go:135  [n1,intExec=create-default-db] Event: "create_database", target: 51, info: {DatabaseName:postgres Statement:CREATE DATABASE IF NOT EXISTS postgres User:root}
I190401 05:35:38.560298 26293 server/server.go:1653  [n1] done ensuring all necessary migrations have run
I190401 05:35:38.560333 26293 server/server.go:1656  [n1] serving sql connections
I190401 05:35:38.562954 26578 sql/event_log.go:135  [n1] Event: "node_join", target: 1, info: {Descriptor:{NodeID:1 Address:127.0.0.1:36033 Attrs: Locality: ServerVersion:2.1-10 BuildTag:v19.1.0-beta.20190318-526-g668162c StartedAt:1554096938449926888 LocalityAddress:[] XXX_NoUnkeyedLiteral:{} XXX_sizecache:0} ClusterID:f23eda40-f846-4ab4-bcc9-88e13cafccbe StartedAt:1554096938449926888 LastUp:1554096938449926888}
I190401 05:35:38.563682 26543 server/server_update.go:67  [n1] no need to upgrade, cluster already at the newest version
I190401 05:35:38.575111 26582 sql/event_log.go:135  [n1,client=127.0.0.1:44984,user=root] Event: "set_cluster_setting", target: 0, info: {SettingName:kv.rangefeed.enabled Value:true User:root}
I190401 05:35:38.579478 26582 sql/event_log.go:135  [n1,client=127.0.0.1:44984,user=root] Event: "set_cluster_setting", target: 0, info: {SettingName:kv.closed_timestamp.target_duration Value:00:00:01 User:root}
I190401 05:35:38.582632 26582 sql/event_log.go:135  [n1,client=127.0.0.1:44984,user=root] Event: "set_cluster_setting", target: 0, info: {SettingName:changefeed.experimental_poll_interval Value:00:00:00.01 User:root}
I190401 05:35:38.585828 26582 sql/event_log.go:135  [n1,client=127.0.0.1:44984,user=root] Event: "create_database", target: 52, info: {DatabaseName:d Statement:CREATE DATABASE d User:root}
I190401 05:35:38.590023 26582 sql/event_log.go:135  [n1,client=127.0.0.1:44984,user=root] Event: "create_table", target: 53, info: {TableName:d.public.foo Statement:CREATE TABLE foo (a INT8 PRIMARY KEY, b STRING) User:root}
I190401 05:35:38.591505 26601 storage/replica_command.go:246  [n1,split,s1,r20/1:/{Table/24-Max}] initiating a split of this range at key /Table/53 [r21] (zone config)
I190401 05:35:38.656546 26293 util/stop/stopper.go:546  quiescing; tasks left:
12     [async] storage.IntentResolver: cleanup txn records
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
1      [async] closedts-rangefeed-subscriber
W190401 05:35:38.656580 26537 storage/intentresolver/intent_resolver.go:820  failed to gc transaction record: could not GC completed transaction anchored at /Table/SystemConfigSpan/Start: node unavailable; try another peer
I190401 05:35:38.656627 26293 util/stop/stopper.go:546  quiescing; tasks left:
11     [async] storage.IntentResolver: cleanup txn records
1      [async] intent_resolver_gc_batcher
1      [async] closedts-rangefeed-subscriber
I190401 05:35:38.656682 26293 util/stop/stopper.go:546  quiescing; tasks left:
3      [async] storage.IntentResolver: cleanup txn records
1      [async] intent_resolver_gc_batcher
1      [async] closedts-rangefeed-subscriber
I190401 05:35:38.656711 26293 util/stop/stopper.go:546  quiescing; tasks left:
2      [async] storage.IntentResolver: cleanup txn records
1      [async] closedts-rangefeed-subscriber
I190401 05:35:38.656744 26293 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] storage.IntentResolver: cleanup txn records
1      [async] closedts-rangefeed-subscriber
I190401 05:35:38.656769 26293 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] closedts-rangefeed-subscriber
    --- FAIL: TestChangefeedDataTTL/sinkless (0.24s)
        changefeed_test.go:1237: expected "must be after replica GC threshold" error got: <nil>

```

SHA: https://github.com/cockroachdb/cockroach/commits/668162cc99e4f3198b663b1abfa51858eeb3ccb8

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestChangefeedDataTTL PKG=github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1213009&tab=buildLog

```
I190401 06:33:22.971356 29212 sql/event_log.go:135  [n1,intExec=initializeClusterSecret] Event: "set_cluster_setting", target: 0, info: {SettingName:cluster.secret Value:85489398-4141-4f30-bba0-1b0caaccf07a User:root}
I190401 06:33:23.012355 29314 sql/event_log.go:135  [n1,intExec=create-default-db] Event: "create_database", target: 50, info: {DatabaseName:defaultdb Statement:CREATE DATABASE IF NOT EXISTS defaultdb User:root}
I190401 06:33:23.038227 29276 sql/event_log.go:135  [n1,intExec=create-default-db] Event: "create_database", target: 51, info: {DatabaseName:postgres Statement:CREATE DATABASE IF NOT EXISTS postgres User:root}
I190401 06:33:23.110241 29061 server/server.go:1653  [n1] done ensuring all necessary migrations have run
I190401 06:33:23.110381 29061 server/server.go:1656  [n1] serving sql connections
I190401 06:33:23.128666 29321 sql/event_log.go:135  [n1] Event: "node_join", target: 1, info: {Descriptor:{NodeID:1 Address:127.0.0.1:34907 Attrs: Locality: ServerVersion:2.1-10 BuildTag:v19.1.0-beta.20190318-526-g668162c StartedAt:1554100402372901894 LocalityAddress:[] XXX_NoUnkeyedLiteral:{} XXX_sizecache:0} ClusterID:faababa7-15cd-49e0-a2dc-bfedc38fc950 StartedAt:1554100402372901894 LastUp:1554100402372901894}
I190401 06:33:23.131093 29318 server/server_update.go:67  [n1] no need to upgrade, cluster already at the newest version
I190401 06:33:23.169224 29054 sql/event_log.go:135  [n1,client=127.0.0.1:57464,user=root] Event: "set_cluster_setting", target: 0, info: {SettingName:kv.rangefeed.enabled Value:true User:root}
I190401 06:33:23.202390 29054 sql/event_log.go:135  [n1,client=127.0.0.1:57464,user=root] Event: "set_cluster_setting", target: 0, info: {SettingName:kv.closed_timestamp.target_duration Value:00:00:01 User:root}
I190401 06:33:23.230782 29054 sql/event_log.go:135  [n1,client=127.0.0.1:57464,user=root] Event: "set_cluster_setting", target: 0, info: {SettingName:changefeed.experimental_poll_interval Value:00:00:00.01 User:root}
I190401 06:33:23.257610 29054 sql/event_log.go:135  [n1,client=127.0.0.1:57464,user=root] Event: "create_database", target: 52, info: {DatabaseName:d Statement:CREATE DATABASE d User:root}
I190401 06:33:23.289914 29054 sql/event_log.go:135  [n1,client=127.0.0.1:57464,user=root] Event: "create_table", target: 53, info: {TableName:d.public.foo Statement:CREATE TABLE foo (a INT8 PRIMARY KEY, b STRING) User:root}
I190401 06:33:23.300002 29248 storage/replica_command.go:246  [n1,split,s1,r20/1:/{Table/24-Max}] initiating a split of this range at key /Table/53 [r21] (zone config)
W190401 06:33:23.526993 29253 storage/intentresolver/intent_resolver.go:820  failed to gc transaction record: could not GC completed transaction anchored at /Table/SystemConfigSpan/Start: node unavailable; try another peer
I190401 06:33:23.526999 29061 util/stop/stopper.go:546  quiescing; tasks left:
12     [async] storage.IntentResolver: cleanup txn records
1      [async] transport racer
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
1      [async] closedts-rangefeed-subscriber
I190401 06:33:23.528276 29061 util/stop/stopper.go:546  quiescing; tasks left:
6      [async] storage.IntentResolver: cleanup txn records
1      [async] transport racer
1      [async] closedts-rangefeed-subscriber
I190401 06:33:23.528546 29061 util/stop/stopper.go:546  quiescing; tasks left:
5      [async] storage.IntentResolver: cleanup txn records
1      [async] transport racer
1      [async] closedts-rangefeed-subscriber
I190401 06:33:23.530048 29061 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] transport racer
1      [async] closedts-rangefeed-subscriber
I190401 06:33:23.531548 29061 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] transport racer
I190401 06:33:23.598782 29234 kv/transport_race.go:113  transport race promotion: ran 58 iterations on up to 301 requests
    --- FAIL: TestChangefeedDataTTL/sinkless (1.39s)
        changefeed_test.go:1237: expected "must be after replica GC threshold" error got: <nil>

```

--- a/pkg/ccl/changefeedccl/cdctest/testfeed.go
+++ b/pkg/ccl/changefeedccl/cdctest/testfeed.go
@@ -50,6 +50,13 @@ type TestFeedMessage struct {
 	Resolved         []byte
 }
 
+func (m TestFeedMessage) String() string {
+	if m.Resolved != nil {
+		return string(m.Resolved)
+	}
+	return fmt.Sprintf(`%s: %s->%s`, m.Topic, m.Key, m.Value)
+}
+
 // TestFeed abstracts over reading from the various types of changefeed sinks.
 type TestFeed interface {
 	// Partitions returns the domain of values that may be returned as a partition

--- a/pkg/ccl/changefeedccl/changefeed_test.go
+++ b/pkg/ccl/changefeedccl/changefeed_test.go
@@ -36,6 +36,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 	"github.com/pkg/errors"
@@ -1173,6 +1174,7 @@ func TestChangefeedDataTTL(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 
 	testFn := func(t *testing.T, db *gosql.DB, f cdctest.TestFeedFactory) {
+		ctx := context.Background()
 		// Set a very simple channel-based, wait-and-resume function as the
 		// BeforeEmitRow hook.
 		var shouldWait int32
@@ -1227,13 +1229,18 @@ func TestChangefeedDataTTL(t *testing.T) {
 		atomic.StoreInt32(&shouldWait, 0)
 		resume <- struct{}{}
 
-		// Verify that the third call to Next() returns an error (the first is the
-		// initial row, the second is the first change. The third should detect the
-		// GC interval mismatch).
-		_, _ = dataExpiredRows.Next()
-		_, _ = dataExpiredRows.Next()
-		if _, err := dataExpiredRows.Next(); !testutils.IsError(err, `must be after replica GC threshold`) {
-			t.Errorf(`expected "must be after replica GC threshold" error got: %+v`, err)
+		// Verify that, at some point, Next() returns a "must be after replica GC
+		// threshold" error. In the common case, that'll be the third call, but
+		// various conditions will cause RangeFeed to emit duplicates and so it may
+		// be a few more.
+		for {
+			msg, err := dataExpiredRows.Next()
+			if testutils.IsError(err, `must be after replica GC threshold`) {
+				break
+			}
+			if msg != nil {
+				log.Infof(ctx, "ignoring message %s", msg)
+			}
 		}
 	}
 

